### PR TITLE
Document Docker Pipe

### PR DESCRIPTION
**Pull request by fkorotkov**
And extract all the possible execution environments in a separate section.

### CODE REVIEW

File: docs/guide/docker-pipe.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+Docker Pipe is a way to execute each [instruction](writing-tasks.md#supported-instructions) in it's own Docker container
+while persisting working directory between each of the containers. For example, you can build your application in 
+one container, run some lint tools in another containers and finally deploy your app via CLI from another container.
+
+**No need to create huge containers with every single tool pre-installed!**
+
+`pipe` can be defined the same way as a `task` with the only difference that [instructions](writing-tasks.md#supported-instructions)
```
Comment (RDIL) [inline docs/guide/docker-pipe.md]:
```suggestion
A `pipe` can be defined the same way as a `task` with the only difference that [instructions](writing-tasks.md#supported-instructions)
```

File: docs/guide/docker-pipe.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+should be grouped under `steps` field defining a Docker `image` for each step to be executed in. Here is an example of how
```
Comment (RDIL) [inline docs/guide/docker-pipe.md]:
```suggestion
should be grouped under the `steps` field defining a Docker `image` for each step to be executed in. Here is an example of how
```

File: docs/guide/docker-pipe.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+we build and validate links of [Cirrus CI documentation](https://github.com/cirruslabs/cirrus-ci-docs) that you are reading right now:
```
Comment (RDIL) [inline docs/guide/docker-pipe.md]:
```suggestion
we build and validate links for the [Cirrus CI documentation](https://github.com/cirruslabs/cirrus-ci-docs) that you are reading right now:
```

File: docs/guide/docker-pipe.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+pipe:
+  name: Build Site and Validate Links
+  steps:
+    - image: squidfunk/mkdocs-material:latest
+      build_script: mkdocs build
+    - image: raviqqe/liche:latest # links validation tool in a separate container
+      validate_script: /liche --document-root=site --recursive site/
+```
+
+Amount of CPU and memory that a pipe is using can be configured with `resources` field:
```
Comment (RDIL) [inline docs/guide/docker-pipe.md]:
```suggestion
Amount of CPU and memory that a pipe has access to can be configured with `resources` field:
```

File: docs/guide/writing-tasks.md [hunk 5]
```diff
@@ -1,63 +1,44 @@
...
-Only one user defined script instruction to run `gradle test` will be executed. Pretty simple, isn't it?
-
-A `task` simply defines a [compute service](supported-computing-services.md) to schedule the task on and 
-a sequence of instructions to execute. The following instructions are supported:
-
-* [`script`](#script-instruction) instruction to execute a script.
-* [`background_script`](#background-script-instruction) instruction to execute a script in a background.
-* [`cache`](#cache-instruction) instruction to persist files between task runs.
-* [`artifacts`](#artifacts-instruction) instruction to store and expose files created via a task.
+The example above defines a single task that will be scheduled and executed on [Linux Community Cluster](linux.md) using `gradle:jdk11` Docker image.
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
The example above defines a single task that will be scheduled and executed on the [Linux Community Cluster](linux.md) using the `gradle:jdk11` Docker image.
```

File: docs/guide/writing-tasks.md [hunk 6]
```diff
@@ -1,63 +1,44 @@
...
+Only one user defined [script instruction](#script-instruction) to run `gradle test` will be executed. Pretty simple, isn't it?
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
Only one user-defined [script instruction](#script-instruction) to run `gradle test` will be executed. Pretty simple, isn't it?
```

File: docs/guide/writing-tasks.md [hunk 7]
```diff
@@ -1,63 +1,44 @@
...
 
 Please read topics below if you want better understand what's doing on in a more complex `.cirrus.yml` configuration file such as this:
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
Please read the topics below if you want better understand what's doing on in a more complex `.cirrus.yml` configuration file, such as this:
```

File: docs/guide/writing-tasks.md [hunk 8]
```diff
@@ -72,7 +53,45 @@ publish_task:
     
     **Note:** instructions withing a task can only be named via a prefix (e.g. `test_script`). 
 
-## Script Instruction
+## Execution Environment
+
+In order to specify where to execute a particular task you can choose from variety of options by defining one of the
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
In order to specify where to execute a particular task you can choose from a variety of options by defining one of the
```

File: docs/guide/writing-tasks.md [hunk 9]
```diff
@@ -72,7 +53,45 @@ publish_task:
...
+following fields of a `task`:
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
following fields for a `task`:
```

File: docs/guide/writing-tasks.md [hunk 10]
```diff
@@ -91,7 +110,7 @@ check_task:
     - gradle check
 ```
 
-## Background Script Instruction
+### Background Script Instruction
 
 A `background_script` instruction is absolutely the same as `script` instruction but Cirrus CI won't wait for the script to finish 
 and will continue execution of following instructions.
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
and will continue execution of further instructions.
```

**Comment by fkorotkov** [comment]
Thank you, @RDIL!

**Comment by RDIL** [comment]
One minor piece of feedback, I think that aliasing `*_pipeline` to `*_pipe` would sound more like traditional CI.

**Comment by fkorotkov** [comment]
@RDIL pipeline usually means having fan-in and fan-out so you can run some parts of it in parallel (like we have `depends_on`). `pipe` name comes from piping in a terminal where you can pipe output of one command as an input of the next command.

I was considering not to introduce a new concept at all and incorporate it into `task` but didn't find a way how it can fit into it. `pipe` is a different concept IMO. The closest thing is [Teckton's Task](https://github.com/tektoncd/pipeline/blob/master/docs/tutorial.md#task).

**Comment by RDIL** [comment]
Ah okay.

**Pull request merged**